### ISSUE TITLE
CS8618 is incorrectly reported for a constructor with SetsRequiredMembersAttribute and a required non-auto property

### ISSUE THREAD

**Issue by vgriph** [dotnet/roslyn#77013]
**Version Used**: 
Dotnet SDK 9.0.101
csc.dll 4.1200.24.57006
C# 13

**Steps to Reproduce**:

1. Create a .NET 9 C# project
2. Add the following class:
  ```cs
    public class Class
    {
        private string @field = string.Empty;

        [SetsRequiredMembers]
        public Class()
        {
        }

        public required string Test
        {
            get => @field;
            set => @field = value;
        }
    }
   ```
4. Compile
 
Also provided as a gist: https://sharplab.io/#gist:c97d05bc1df387299a3a36043bfcd68f
 
**Diagnostic Id**:
CS8618: Non-nullable property 'Test' must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring the property as nullable.

**Expected Behavior**:

The code should compile without warning

**Actual Behavior**:
The code gets a warning suggesting to add the required modifier, when it is already there

**Comment by RikkiGibson** [comment]
This is the intended guardrail for `[SetsRequiredMembers]`. The constructor receiving the warning has an obligation to set that property, which it is not honoring. We don't do a full-on assignment analysis to see if you set the property, but, if it's non-nullable, it had better be put into the non-nullable state.

@333fred in case he wishes to add further information.

**Comment by vgriph** [comment]
I still think the warning message is wrong. It states "Consider adding the 'required' modifier or declaring the property as nullable." But it already has the required modifier.

The example provided is simplified from my actual usage. Where the setter does some calculations to update the backing field. I want the property to be required for other constructors, but in a deserialization constructor I want to avoid having to do the calculations and instead just set the backing field directly.

**Comment by RikkiGibson** [comment]
We could consider introducing a separate diagnostic message when the property is already marked 'required', which doesn't suggest to add that modifier. If you feel strongly enough about it, I would review a PR which adds that message.

> in a deserialization constructor I want to avoid having to do the calculations and instead just set the backing field directly.

I think that if that constructor does not assign the property anywhere, it makes the most sense to wrap the constructor signature in `#pragma warning disable/restore CS8618`. This expresses that the requirement is being handled in some way the compiler is unable to track.
